lib/smart_answer/state_resolver.rb

```diff
@@ -0,0 +1,78 @@
+module SmartAnswer
+  class StateResolver
+    def initialize(flow)
+      @flow = flow
+    end
+
+    def state_from_response_store(response_store, requested_node = nil)
+      forwarding_responses = @flow.response_store == :session ? {} : response_store.all
+      state = start_state(forwarding_responses)
+      loop do
+        node_name = state.current_node.to_s
+        return state unless response_store.has?(node_name)
+
+        response = response_store.get(node_name)
+        return apply_response_to_state(state, response) if node_name == requested_node
+
+        new_state = transition_state_to_next_node(state, response)
+        return new_state if new_state.error
+        return new_state if @flow.node(new_state.current_node).outcome?
+
+        state = new_state
+      end
+    end
+
+    def state_from_params(params)
+      responses = params[:responses].to_s.split("/")
+
+      state = responses.inject(start_state) do |current_state, response|
+        return current_state if current_state.error
+
+        transition_state_to_next_node(current_state, response)
+      end
+
+      if params[:next]
+        transition_state_to_next_node(state, params[:response])
+      elsif params[:previous_response]
+        apply_response_to_state(state, params[:previous_response])
+      else
+        state
+      end
+    end
+
+  private
+
+    def start_state(forwarding_responses = {})
+      State.new(@flow.questions.first.name,
+                forwarding_responses: forwarding_responses).freeze
+    end
+
+    def apply_response_to_state(state, response)
+      # check for errors in the response by seeing if we can reach next state
+      @flow.node(state.current_node).transition(state, response)
```

Comment:
Calling that would cause a different behaviour unfortunately as `transition_state_to_next_node` catches the exception. So we'd have to set a variable for the response of transition as one state and add a conditional to see if that state is used or the other (original) state is modified. So I think it'd end up a bit more complex to read eventually.

Reply:
I've given this a re-write to try make it all clearer. I figured rather than answering your question by explaining the code it'd be better to try have the code it self easier to understand. I had to add another comment to it but I think it benefits from it as it's rather quirky.